In [1]:
import pandas as pd
import numpy as np
from sklearn.naive_bayes import BernoulliNB
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.metrics import confusion_matrix

In [2]:
df = pd.read_csv('/home/vivek/Downloads/sentiment labelled sentences/imdb_labelled.txt', delimiter='\t', header=None)
df.columns = ['reviews', 'label']

In [3]:
df.head()

reviews  label
0  A very, very, very slow-moving, aimless movie ...      0
1  Not sure who was more lost - the flat characte...      0
2  Attempting artiness with black & white and cle...      0
3       Very little music or anything to speak of.        0
4  The best scene in the movie was when Gerardo i...      1

In [4]:
negative_words1 = ['lose', 'loss', 'kill', 'bad', 'hate']

In [5]:
negative_words2 = ['lose', 'loss', 'kill', 'bad', 'hate', 'fake', 'poor', 'lost', 'low', 'cut']

In [6]:
negative_words3 = ['lose', 'loss', 'kill', 'bad', 'hate', 'fake', 'poor', 'lost', 'low', 'cut', 'worst', 'slow', 'little', 'ridiculous', 'disappoint']

In [7]:
negative_words4 = ['lose', 'loss', 'kill', 'bad', 'hate', 'fake', 'poor', 'lost', 'low', 'cut', 'worst', 'slow', 'little', 'ridiculous', 'disappoint', 'unhappy', 'sad', 'sore', 'alone', 'dead']

In [8]:
negative_words5 = ['lose', 'loss', 'kill', 'bad', 'hate', 'fake', 'poor', 'lost', 'low', 'cut', 'worst', 'slow', 'little', 'ridiculous', 'disappoint', 'unhappy', 'sad', 'sore', 'alone', 'dead', 'tired', 'cry', 'lie', 'small', 'narrow']

We have created 5 different types of feature lists with increment of 5 features in each. Next, we will make different classifiers using these lists and check their accuracy.

In [9]:
for word in negative_words1:
    df[str(word)] = df['reviews'].str.contains(str(word), case=False)

In [10]:
x_train, x_test, y_train, y_test = train_test_split(df[negative_words1], df['label'], test_size=0.2, random_state=20)

In [11]:
bnb1 = BernoulliNB()

In [12]:
bnb1.fit(x_train, y_train).score(x_test, y_test)

0.5666666666666667

In [13]:
confusion_matrix(bnb1.predict(x_train), y_train)

array([[ 52,   6],
       [238, 302]])

**False positive** here is 6 and **False Negative** is 238. So, our model is not able to predict for bad review.

In [14]:
for word in negative_words2:
    df[str(word)] = df['reviews'].str.contains(str(word), case=False)

In [15]:
x_train, x_test, y_train, y_test = train_test_split(df[negative_words2], df['label'], test_size=0.2, random_state=20)

In [16]:
bnb2 = BernoulliNB()

In [17]:
bnb2.fit(x_train, y_train).score(x_test, y_test)

0.56

In [18]:
confusion_matrix(bnb2.predict(x_train), y_train)

array([[ 74,  14],
       [216, 294]])

In [19]:
for word in negative_words3:
    df[str(word)] = df['reviews'].str.contains(str(word), case=False)

In [20]:
x_train, x_test, y_train, y_test = train_test_split(df[negative_words3], df['label'], test_size=0.2, random_state=20)

In [21]:
bnb3 = BernoulliNB()

In [22]:
bnb3.fit(x_train, y_train).score(x_test, y_test)

0.5933333333333334

In [23]:
confusion_matrix(bnb3.predict(x_train), y_train)

array([[ 92,  17],
       [198, 291]])

In [24]:
for word in negative_words4:
    df[str(word)] = df['reviews'].str.contains(str(word), case=False)

In [25]:
x_train, x_test, y_train, y_test = train_test_split(df[negative_words4], df['label'], test_size=0.2, random_state=20)

In [26]:
bnb4 = BernoulliNB()

In [27]:
bnb4.fit(x_train, y_train).score(x_test, y_test)

0.5933333333333334

In [28]:
confusion_matrix(bnb4.predict(x_train), y_train)

array([[ 92,  17],
       [198, 291]])

In [29]:
for word in negative_words5:
    df[str(word)] = df['reviews'].str.contains(str(word), case=False)

In [30]:
x_train, x_test, y_train, y_test = train_test_split(df[negative_words5], df['label'], test_size=0.2, random_state=20)

In [31]:
bnb5 = BernoulliNB()

In [32]:
bnb5.fit(x_train, y_train).score(x_test, y_test)

0.5733333333333334

In [33]:
confusion_matrix(bnb5.predict(x_train), y_train)

array([[102,  26],
       [188, 282]])

In [34]:
cross_val_score(bnb1, df[negative_words1], df['label'], cv=10)

array([0.55263158, 0.56578947, 0.56      , 0.62666667, 0.64      ,
       0.62666667, 0.51351351, 0.59459459, 0.52702703, 0.66216216])

In [35]:
cross_val_score(bnb2, df[negative_words2], df['label'], cv=10)

array([0.64473684, 0.57894737, 0.58666667, 0.62666667, 0.64      ,
       0.61333333, 0.54054054, 0.63513514, 0.54054054, 0.62162162])

In [36]:
cross_val_score(bnb3, df[negative_words3], df['label'], cv=10)

array([0.67105263, 0.61842105, 0.57333333, 0.62666667, 0.68      ,
       0.65333333, 0.56756757, 0.64864865, 0.60810811, 0.63513514])

In [37]:
cross_val_score(bnb4, df[negative_words4], df['label'], cv=10)

array([0.67105263, 0.61842105, 0.57333333, 0.62666667, 0.68      ,
       0.65333333, 0.56756757, 0.64864865, 0.60810811, 0.63513514])

In [38]:
cross_val_score(bnb5, df[negative_words5], df['label'], cv=10)

array([0.67105263, 0.61842105, 0.57333333, 0.62666667, 0.68      ,
       0.65333333, 0.56756757, 0.64864865, 0.56756757, 0.63513514])

**From above 5 cross validation score for different classifiers, it is clear that from 3rd classifier there is no effect of adding new feature or may be those features are not relevent with our data.**

If we look into our confusion matrix for every classifier then **bnb5** is performing best. Though, it's score says it is overfitted.